[X86] Lowering of load atomic vector of float via cast

jofrn opened the pull request:
X86 backend does not lower load atomic \<n x float\>; yet it does so for integers, so we can cast to an integer before lowering in order to support this instruction.

jofrn committed:
[X86] Lowering of load atomic float via cast

X86 backend does not lower load atomic float, so it can be casted to an
integer before lowering.

arsenm commented on llvm/test/CodeGen/X86/atomic-float.ll:
```diff
@@ -0,0 +1,15 @@
+; CHECK-NEXT:    [[TMP2:%.*]] = bitcast i32 [[TMP1]] to float
+; CHECK-NEXT:    ret float [[TMP2]]
+;
+  %src = alloca float
```
Use a pointer argument instead of an alloca

arsenm commented on llvm/test/CodeGen/X86/atomic-float.ll:
```diff
@@ -0,0 +1,15 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s --mtriple=x86_64 --passes=atomic-expand -S -o - | FileCheck %s
```
This belongs in test/Transforms/AtomicExpand/X86

arsenm replied:
Are there missing codegen tests for this? Those should be added if they are missing

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -30976,6 +30976,14 @@ bool X86TargetLowering::needsCmpXchgNb(Type *MemType) const {
 
+TargetLoweringBase::AtomicExpansionKind
+X86TargetLowering::shouldCastAtomicLoadInIR(LoadInst *LI) const {
+  if (LI->getType()->isVectorTy())
+    if (cast<VectorType>(LI->getType())->getElementType()->isFloatingPointTy())
```
isVectorTy is redundant with the cast. Use dyn_cast

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -30976,6 +30976,14 @@ bool X86TargetLowering::needsCmpXchgNb(Type *MemType) const {
   return false;
 }
 
+TargetLoweringBase::AtomicExpansionKind
```
I would like to get rid of this hook. Directly promoting atomic_load to int in codegen, like any other operation, shouldn't be difficult

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -30976,6 +30976,14 @@ bool X86TargetLowering::needsCmpXchgNb(Type *MemType) const {
 
+TargetLoweringBase::AtomicExpansionKind
+X86TargetLowering::shouldCastAtomicLoadInIR(LoadInst *LI) const {
+  if (LI->getType()->isVectorTy())
```
Add a test case for vector type?

arsenm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -137,6 +137,10 @@ static cl::opt<bool> VerifyNoAliasScopeDomination(
     cl::desc("Ensure that llvm.experimental.noalias.scope.decl for identical "
              "scopes are not dominating"));
 
+static cl::opt<bool> AllowAtomicVector(
```
This patch should not touch the verifier, and there should never be an option to make IR valid or not

jofrn committed:
[Verifier] Allow vector type in atomic load and store

Vector types on atomics are assumed to be invalid by the verifier. However,
this type can be valid if it is lowered by codegen.

shiltian commented:
it seems like this PR is stacked but not properly stacked

arsenm commented on llvm/test/CodeGen/X86/atomicvec-float.ll:
```diff
@@ -0,0 +1,38 @@
+; CHECK-LABEL: load_atomic_vector_float1:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    movss {{.*#+}} xmm0 = mem[0],zero,zero,zero
+; CHECK-NEXT:    retq
```
Focus on one thing at a time, only handle scalar FP cases in this PR

jofrn replied:
Sorry. Meant to have title reflect vectors only initially. This is stacked on top of the IR/Verifier change to permit vectors in atomic load and stores.

arsenm commented on llvm/test/CodeGen/X86/atomicvec-float.ll:
```diff
@@ -0,0 +1,38 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s --mtriple=x86_64 | FileCheck %s
+
+define float @load_atomic_float(ptr %src) {
```
Test all the FP types, float, double, half, bfloat. fp128 if 128-bit atomics work

arsenm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -4255,9 +4255,10 @@ void Verifier::visitLoadInst(LoadInst &LI) {
-    Check(ElTy->isIntOrPtrTy() || ElTy->isFloatingPointTy(),
-          "atomic load operand must have integer, pointer, or floating point "
-          "type!",
+    Check(ElTy->getScalarType()->isIntOrPtrTy() ||
```
This PR shouldn't touch the verifier, only handle codegen for atomic load / store

jofrn replied:
It is stacked on top of https://github.com/llvm/llvm-project/pull/117625. Please take a look. This PR handles vectors only -- renamed the title.

arsenm commented on llvm/test/CodeGen/X86/atomicvec-float.ll:
```diff
@@ -0,0 +1,38 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s --mtriple=x86_64 | FileCheck %s
```
There is already a deceptively named atomic-fp.ll. llvm/test/CodeGen/X86/atomic-non-integer.ll seems like it already exists?

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -31158,6 +31158,14 @@ bool X86TargetLowering::needsCmpXchgNb(Type *MemType) const {
   return false;
 }
 
+TargetLoweringBase::AtomicExpansionKind
+X86TargetLowering::shouldCastAtomicLoadInIR(LoadInst *LI) const {
+  if (const auto VT = dyn_cast<VectorType>(LI->getType()))
+    if (VT->getElementType()->isFloatingPointTy())
+      return AtomicExpansionKind::CastToInteger;
+  return TargetLowering::shouldCastAtomicLoadInIR(LI);
+}
```
This seems to be moving in the opposite direction. The existing tests atomic FP load already works, and this is regressing it to use the IR expansion (which should be deleted)

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -31158,6 +31158,14 @@ bool X86TargetLowering::needsCmpXchgNb(Type *MemType) const {
 
+TargetLoweringBase::AtomicExpansionKind
+X86TargetLowering::shouldCastAtomicLoadInIR(LoadInst *LI) const {
+  if (const auto VT = dyn_cast<VectorType>(LI->getType()))
+    if (VT->getElementType()->isFloatingPointTy())
+      return AtomicExpansionKind::CastToInteger;
```
Title is also wrong, the point is to handle vectors?

jofrn committed:
Add tests for double,half,bfloat,fp128

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -31158,6 +31158,14 @@ bool X86TargetLowering::needsCmpXchgNb(Type *MemType) const {
 
+TargetLoweringBase::AtomicExpansionKind
+X86TargetLowering::shouldCastAtomicLoadInIR(LoadInst *LI) const {
+  if (const auto VT = dyn_cast<VectorType>(LI->getType()))
+    if (VT->getElementType()->isFloatingPointTy())
```
Should move this into the default implementation. Targets can opt out as the hook is removed
